osu.Game/Rulesets/Objects/HitObject.cs

```diff
@@ -215,6 +216,34 @@ public HitSampleInfo CreateHitSampleInfo(string sampleName = HitSampleInfo.HIT_N
 
             return new HitSampleInfo(sampleName);
         }
+
+        public HitObject DeepClone(IDictionary<object, object> referenceLookup)
+        {
+            if (referenceLookup.TryGetValue(this, out object existing))
+                return (HitObject)existing;
+
+            var clone = CreateInstance();
+            referenceLookup[this] = clone;
+
+            clone.CopyFrom(this, referenceLookup);
+
+            return clone;
+        }
+
+        protected virtual HitObject CreateInstance() => new HitObject();
```

Comment:
Is there a reason we can't use 

```suggestion
        protected HitObject CreateInstance() => Activator.CreateInstance(GetType())!;
```

Reply:
certain subtypes do not provide parameterless constructors or otherwise require some consideration about how they get constructed during cloning, so this tends to be more straightforward.

Reply:
I see. For reference, here's a few examples:

`SliderEndCircle`
`StrongNestedHit`

Reply:
To fix tests (and maybe relieve the overhead from the end user(, the base implementation could be left `virtual` but with a default implementation as I propose. This would fix the countless cases of `TestHitObject` – and also make a lot of the manual implementations redundant. To handle the edge cases, it could be something like

```csharp
        protected virtual HitObject CreateInstance()
        {
            try
            {
                return (HitObject)Activator.CreateInstance(GetType());
            }
            catch
            {
                throw new NotImplementedException($"Hit object of type {GetType().ReadableName()} requires constructor parameters but does not override {nameof(CreateInstance)}.");
            }
        }
```

Reply:
I would like to avoid reflection altogether because it's not easily AOT-able.